gapis/atom/group.go

```diff
@@ -311,11 +338,29 @@ func (g *Group) AddGroup(start, end ID, name string) error {
 	return nil
 }
 
+func (c *SubcommandRange) Insert(r SubRange) {
+	var lastRange *SubcommandRange
+	if len(c.SubRanges) > 0 {
+		lastRange = c.SubRanges[len(c.SubRanges)-1]
+	}
+
+	if lastRange == nil ||
+		lastRange.QualifiedIndex[len(lastRange.QualifiedIndex)-1] != r[0] {
+		lastRange = NewRange(Range{ID(r[0]), ID(r[0]) + 1}, c.QualifiedIndex)
+		lastRange.QualifiedIndex = append(lastRange.QualifiedIndex, r[0])
+		c.SubRanges = append(c.SubRanges, lastRange)
+	}
+	if len(r) > 1 {
+		lastRange.Insert(r[1:])
+	}
+}
+
 // AddAtoms fills the group and sub-groups with atoms based on the predicate
 // pred. If maxChildren is positive, the group, and any of it's decendent
 // groups, which have more than maxChildren child elements, will have their
 // children grouped into new synthetic groups of at most maxChildren children.
-func (g *Group) AddAtoms(pred func(id ID) bool, maxChildren uint64) error {
+func (g *Group) AddAtoms(pred func(id ID) bool, maxChildren uint64,
```

Comment:
The `sr` parameter is a bit gross. It would be nicer if `AddAtoms` could automatically traverse into subcommands (requires building before calling `AddAtoms`), or maybe some sort of callback passed to `pred` for declaring subcommands? I prefer the former.